# Introduction

Après avoir parcouru le fichier, nous allons commencer à le modifier

# Import des modules

In [1]:
import pandas as pd

# Lecture des données

In [2]:
df = pd.read_csv("datas.csv", index_col=0)

# Quelques rappels

### 10 première lignes

In [3]:
df.head(10)

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
Vol_0  2006      5           8          1    705.0         700    917.0   
Vol_1  2006      5          10          3    705.0         700    920.0   
Vol_2  2006      5          12          5    700.0         700    922.0   
Vol_3  2006      5          14          7    700.0         700    910.0   
Vol_4  2006      5          16          2    724.0         700    921.0   
Vol_5  2006      5          18          4    658.0         700    909.0   
Vol_6  2006      5          20          6    657.0         700    908.0   
Vol_7  2006      5          22          1    653.0         700    904.0   
Vol_8  2006      5          24          3    658.0         700    914.0   
Vol_9  2006      5          26          5    800.0         700   1017.0   

       CRSArrTime UniqueCarrier  FlightNum        ...         TaxiIn  TaxiOut  \
Vol_0         915            US        560        ...              6        9   
Vol_1         915            US        560        ...              5       10   
Vol_2         915            US        560        ...              8       11   
Vol_3         915            US        560        ...              5        5   
Vol_4         915            US        560        ...              3        9   
Vol_5         915            US        560        ...              6       12   
Vol_6         915            US        560        ...              3       11   
Vol_7         915            US        560        ...              3        9   
Vol_8         915            US        560        ...              4        9   
Vol_9         915            US        560        ...              9       11   

       Cancelled  CancellationCode  Diverted  CarrierDelay WeatherDelay  \
Vol_0          0               NaN         0             0            0   
Vol_1          0               NaN         0             0            0   
Vol_2          0               NaN         0             0            0   
Vol_3          0               NaN         0             0            0   
Vol_4          0               NaN         0             0            0   
Vol_5          0               NaN         0             0            0   
Vol_6          0               NaN         0             0            0   
Vol_7          0               NaN         0             0            0   
Vol_8          0               NaN         0             0            0   
Vol_9          0               NaN         0            60            0   

      NASDelay  SecurityDelay  LateAircraftDelay  
Vol_0        0              0                  0  
Vol_1        0              0                  0  
Vol_2        0              0                  0  
Vol_3        0              0                  0  
Vol_4        0              0                  0  
Vol_5        0              0                  0  
Vol_6        0              0                  0  
Vol_7        0              0                  0  
Vol_8        0              0                  0  
Vol_9        2              0                  0  

[10 rows x 29 columns]

### 10 dernière lignes

In [4]:
df.tail(10)

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
Vol_301450  2006      5          31          3   1922.0        1858   2151.0   
Vol_301451  2006      5          31          3   1501.0        1504   1841.0   
Vol_301452  2006      5          31          3   1851.0        1850   2024.0   
Vol_301453  2006      5          31          3    640.0         645   1408.0   
Vol_301454  2006      5          31          3   1151.0        1020   1833.0   
Vol_301455  2006      5          31          3   1838.0        1842   1943.0   
Vol_301456  2006      5          31          3    640.0         645   1307.0   
Vol_301457  2006      5          31          3   1222.0        1155   1422.0   
Vol_301458  2006      5          31          3   1830.0        1800   2057.0   
Vol_301459  2006      5          31          3   1615.0        1615   1755.0   

            CRSArrTime UniqueCarrier  FlightNum        ...         TaxiIn  \
Vol_301450        2122            DL        371        ...              7   
Vol_301451        1826            DL        373        ...              7   
Vol_301452        2033            DL        376        ...              6   
Vol_301453        1400            DL        378        ...              8   
Vol_301454        1710            DL        379        ...             11   
Vol_301455        1945            DL        384        ...             12   
Vol_301456        1318            DL        385        ...              5   
Vol_301457        1345            DL        387        ...             11   
Vol_301458        2037            DL        389        ...              9   
Vol_301459        1804            DL        392        ...              6   

            TaxiOut  Cancelled  CancellationCode  Diverted  CarrierDelay  \
Vol_301450       12          0               NaN         0            24   
Vol_301451       40          0               NaN         0             0   
Vol_301452       18          0               NaN         0             0   
Vol_301453       41          0               NaN         0             0   
Vol_301454       16          0               NaN         0            83   
Vol_301455       16          0               NaN         0             0   
Vol_301456       13          0               NaN         0             0   
Vol_301457       24          0               NaN         0            27   
Vol_301458       25          0               NaN         0             0   
Vol_301459       14          0               NaN         0             0   

           WeatherDelay NASDelay  SecurityDelay  LateAircraftDelay  
Vol_301450            0        5              0                  0  
Vol_301451            0       15              0                  0  
Vol_301452            0        0              0                  0  
Vol_301453            0        0              0                  0  
Vol_301454            0        0              0                  0  
Vol_301455            0        0              0                  0  
Vol_301456            0        0              0                  0  
Vol_301457            0       10              0                  0  
Vol_301458            0       17              0                  3  
Vol_301459            0        0              0                  0  

[10 rows x 29 columns]

### Liste des colonnes

In [5]:
df.columns

Index(['Year', 'Month', 'DayofMonth', 'DayOfWeek', 'DepTime', 'CRSDepTime',
       'ArrTime', 'CRSArrTime', 'UniqueCarrier', 'FlightNum', 'TailNum',
       'ActualElapsedTime', 'CRSElapsedTime', 'AirTime', 'ArrDelay',
       'DepDelay', 'Origin', 'Dest', 'Distance', 'TaxiIn', 'TaxiOut',
       'Cancelled', 'CancellationCode', 'Diverted', 'CarrierDelay',
       'WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay'],
      dtype='object')

### Dimensions du tableau

In [6]:
df.shape

(301460, 29)

### Type des données

In [7]:
df.dtypes

Year                   int64
Month                  int64
DayofMonth             int64
DayOfWeek              int64
DepTime              float64
CRSDepTime             int64
ArrTime              float64
CRSArrTime             int64
UniqueCarrier         object
FlightNum              int64
TailNum               object
ActualElapsedTime    float64
CRSElapsedTime       float64
AirTime              float64
ArrDelay             float64
DepDelay             float64
Origin                object
Dest                  object
Distance               int64
TaxiIn                 int64
TaxiOut                int64
Cancelled              int64
CancellationCode      object
Diverted               int64
CarrierDelay           int64
WeatherDelay           int64
NASDelay               int64
SecurityDelay          int64
LateAircraftDelay      int64
dtype: object

### Décompte des valeurs présentes

In [8]:
df.count()

Year                 301460
Month                301460
DayofMonth           301460
DayOfWeek            301460
DepTime              297914
CRSDepTime           301460
ArrTime              297324
CRSArrTime           301460
UniqueCarrier        301460
FlightNum            301460
TailNum              301460
ActualElapsedTime    297324
CRSElapsedTime       301460
AirTime              297324
ArrDelay             297324
DepDelay             297914
Origin               301460
Dest                 301460
Distance             301460
TaxiIn               301460
TaxiOut              301460
Cancelled            301460
CancellationCode       3546
Diverted             301460
CarrierDelay         301460
WeatherDelay         301460
NASDelay             301460
SecurityDelay        301460
LateAircraftDelay    301460
dtype: int64

### Première infos statistiques

In [9]:
df.describe()

Year     Month     DayofMonth      DayOfWeek        DepTime  \
count  301460.0  301460.0  301460.000000  301460.000000  297914.000000   
mean     2006.0       5.0      15.928073       3.743601    1338.087599   
std         0.0       0.0       8.946855       1.985304     473.382507   
min      2006.0       5.0       1.000000       1.000000       1.000000   
25%      2006.0       5.0       8.000000       2.000000     930.000000   
50%      2006.0       5.0      16.000000       4.000000    1327.000000   
75%      2006.0       5.0      24.000000       5.000000    1730.000000   
max      2006.0       5.0      31.000000       7.000000    2431.000000   

          CRSDepTime        ArrTime     CRSArrTime      FlightNum  \
count  301460.000000  297324.000000  301460.000000  301460.000000   
mean     1330.788725    1486.539930    1494.931002    2207.965889   
std       460.417894     496.297795     476.034379    2025.136577   
min        10.000000       1.000000       1.000000       1.000000   
25%       930.000000    1109.000000    1115.000000     576.000000   
50%      1323.000000    1516.000000    1520.000000    1476.000000   
75%      1720.000000    1909.000000    1905.000000    3640.000000   
max      2359.000000    2550.000000    2359.000000    9501.000000   

       ActualElapsedTime        ...               Distance         TaxiIn  \
count      297324.000000        ...          301460.000000  301460.000000   
mean          125.134967        ...             730.104979       6.797406   
std            71.050381        ...             577.823534      29.621836   
min            16.000000        ...              24.000000       0.000000   
25%            74.000000        ...             316.000000       4.000000   
50%           106.000000        ...             576.000000       5.000000   
75%           155.000000        ...             957.000000       7.000000   
max          1511.000000        ...            4962.000000    1467.000000   

             TaxiOut      Cancelled       Diverted   CarrierDelay  \
count  301460.000000  301460.000000  301460.000000  301460.000000   
mean       15.426965       0.011763       0.001957       2.879888   
std        10.887856       0.107817       0.044196      17.168730   
min         0.000000       0.000000       0.000000       0.000000   
25%        10.000000       0.000000       0.000000       0.000000   
50%        13.000000       0.000000       0.000000       0.000000   
75%        18.000000       0.000000       0.000000       0.000000   
max       292.000000       1.000000       1.000000    1064.000000   

        WeatherDelay       NASDelay  SecurityDelay  LateAircraftDelay  
count  301460.000000  301460.000000  301460.000000      301460.000000  
mean        0.541531       3.194497       0.014264           4.011441  
std         7.330082      14.445620       0.799616          18.106906  
min         0.000000       0.000000       0.000000           0.000000  
25%         0.000000       0.000000       0.000000           0.000000  
50%         0.000000       0.000000       0.000000           0.000000  
75%         0.000000       0.000000       0.000000           0.000000  
max       646.000000     495.000000     179.000000         515.000000  

[8 rows x 24 columns]

### Mémoire occupée par le DataFrame

In [13]:
df.memory_usage() / 1024 / 1024

Index                2.299957
Year                 2.299957
Month                2.299957
DayofMonth           2.299957
DayOfWeek            2.299957
DepTime              2.299957
CRSDepTime           2.299957
ArrTime              2.299957
CRSArrTime           2.299957
UniqueCarrier        2.299957
FlightNum            2.299957
TailNum              2.299957
ActualElapsedTime    2.299957
CRSElapsedTime       2.299957
AirTime              2.299957
ArrDelay             2.299957
DepDelay             2.299957
Origin               2.299957
Dest                 2.299957
Distance             2.299957
TaxiIn               2.299957
TaxiOut              2.299957
Cancelled            2.299957
CancellationCode     2.299957
Diverted             2.299957
CarrierDelay         2.299957
WeatherDelay         2.299957
NASDelay             2.299957
SecurityDelay        2.299957
LateAircraftDelay    2.299957
dtype: float64

In [14]:
df.memory_usage().sum() / 1024 / 1024

68.99871826171875

# Suppression conditionnelle de lignes
Les vols déviés de leurs plans d'origine sont très rares et ne nous intéressent pas ici. on choisit donc de les retirer de notre DataFrame. On utilise pour cela l'extration conditionnelle

In [18]:
df.Diverted.value_counts()

0    300870
Name: Diverted, dtype: int64

In [19]:
df = df[df.Diverted == 0]

In [20]:
df.shape

(300870, 29)

### Exercice : faire la même chose avec les vols annulés

In [23]:
df.Cancelled.value_counts()

0    297324
Name: Cancelled, dtype: int64

In [21]:
df = df[df.Cancelled == 0]

In [22]:
df.shape

(297324, 29)

# Suppression de colonnes
Certaines colonnes sont devenues inutiles du fait de notre extraction conditionnelle précédente

### Exercice : supprimer toutes les colonnes inutiles ainsi que taxiIn et taxiOut

In [25]:
df.drop('Diverted', axis=1, inplace=True)

In [26]:
df.drop(['Cancelled', 'CancellationCode'], axis=1, inplace=True)

In [27]:
df.drop('Year', axis=1, inplace=True)

In [31]:
df.drop('Month', axis=1, inplace=True)

In [28]:
df.drop(['TaxiIn', 'TaxiOut'], axis=1, inplace=True)

In [34]:
df.head()

DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime  \
Vol_0           8          1    705.0         700    917.0         915   
Vol_1          10          3    705.0         700    920.0         915   
Vol_2          12          5    700.0         700    922.0         915   
Vol_3          14          7    700.0         700    910.0         915   
Vol_4          16          2    724.0         700    921.0         915   

      UniqueCarrier  FlightNum TailNum  ActualElapsedTime        ...          \
Vol_0            US        560  N405US              132.0        ...           
Vol_1            US        560  N784AU              135.0        ...           
Vol_2            US        560  N421US              142.0        ...           
Vol_3            US        560  N435US              130.0        ...           
Vol_4            US        560  N426US              117.0        ...           

       ArrDelay  DepDelay  Origin  Dest Distance CarrierDelay  WeatherDelay  \
Vol_0       2.0       5.0     PBI   DCA      857            0             0   
Vol_1       5.0       5.0     PBI   DCA      857            0             0   
Vol_2       7.0       0.0     PBI   DCA      857            0             0   
Vol_3      -5.0       0.0     PBI   DCA      857            0             0   
Vol_4       6.0      24.0     PBI   DCA      857            0             0   

       NASDelay  SecurityDelay  LateAircraftDelay  
Vol_0         0              0                  0  
Vol_1         0              0                  0  
Vol_2         0              0                  0  
Vol_3         0              0                  0  
Vol_4         0              0                  0  

[5 rows x 22 columns]

# Mémoire occupée
On regarde l'évolution de la mémoire occupée par le DataFrame suite à la suppression de plusieurs colonnes

In [32]:
df.memory_usage().sum() / 1024 / 1024

52.173248291015625

# Optimisation des types de variable
La méthode astype nous permet de changer le dtype d'une variable, ceci afin d'économiser de la mémoire et potentiellement accélérer certaines opérations

### exemple de dayOfMonth

In [36]:
df.DayofMonth.memory_usage() / 1024 / 1024

4.53680419921875

In [37]:
print(df.DayofMonth.min(), df.DayofMonth.max())

1 31


In [38]:
df.DayofMonth = df.DayofMonth.astype('int8')

In [39]:
print(df.DayofMonth.head(), df.DayofMonth.tail())

Vol_0     8
Vol_1    10
Vol_2    12
Vol_3    14
Vol_4    16
Name: DayofMonth, dtype: int8 Vol_301455    31
Vol_301456    31
Vol_301457    31
Vol_301458    31
Vol_301459    31
Name: DayofMonth, dtype: int8


In [40]:
df.DayofMonth.memory_usage() / 1024 / 1024

2.551952362060547

### appliquer à DayOfWeek

In [41]:
df.DayOfWeek.memory_usage() / 1024 / 1024

4.53680419921875

In [42]:
print(df.DayOfWeek.min(), df.DayOfWeek.max())

1 7


In [43]:
df.DayOfWeek = df.DayOfWeek.astype('int8')

In [44]:
print(df.DayOfWeek.head(), df.DayOfWeek.tail())

Vol_0    1
Vol_1    3
Vol_2    5
Vol_3    7
Vol_4    2
Name: DayOfWeek, dtype: int8 Vol_301455    3
Vol_301456    3
Vol_301457    3
Vol_301458    3
Vol_301459    3
Name: DayOfWeek, dtype: int8


In [45]:
df.DayOfWeek.memory_usage() / 1024 / 1024

2.551952362060547

### appliquer à DepTime, ArrTime, CRSDepTime et CRSArrTime (attention, il y a un piège !)

In [46]:
df.DepTime.memory_usage() / 1024 / 1024

4.53680419921875

In [47]:
print(df.DepTime.min(), df.DepTime.max())

1.0 2431.0


In [48]:
df.DepTime = df.DepTime.astype('int16')

In [49]:
print(df.DepTime.head(), df.DepTime.tail())

Vol_0    705
Vol_1    705
Vol_2    700
Vol_3    700
Vol_4    724
Name: DepTime, dtype: int16 Vol_301455    1838
Vol_301456     640
Vol_301457    1222
Vol_301458    1830
Vol_301459    1615
Name: DepTime, dtype: int16


In [50]:
df.DepTime.memory_usage() / 1024 / 1024

2.8355026245117188

In [51]:
df.ArrTime = df.ArrTime.astype('int16')

In [52]:
df.CRSDepTime = df.CRSDepTime.astype('int16')

In [53]:
df.CRSArrTime = df.CRSArrTime.astype('int16')

# Mémoire occupée
On regarde une dernière fois la mémoire occupée par le DataFrame suite aux changements de dtype

In [54]:
df.memory_usage().sum() / 1024 / 1024

41.398338317871094

# Création de colonnes
On souhaite créer une colonne renseignant, à partir des différents types de retard, le retard total de chaque vol

### Exercice : Chercher comment créer la colonne du retard total (3 méthodes possibles)

### méthode 1

In [56]:
df['TotalDelay'] = df.CarrierDelay + df.WeatherDelay + df.NASDelay + df.SecurityDelay + df.LateAircraftDelay

In [57]:
df.head()

DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime  \
Vol_0           8          1      705         700      917         915   
Vol_1          10          3      705         700      920         915   
Vol_2          12          5      700         700      922         915   
Vol_3          14          7      700         700      910         915   
Vol_4          16          2      724         700      921         915   

      UniqueCarrier  FlightNum TailNum  ActualElapsedTime     ...      \
Vol_0            US        560  N405US              132.0     ...       
Vol_1            US        560  N784AU              135.0     ...       
Vol_2            US        560  N421US              142.0     ...       
Vol_3            US        560  N435US              130.0     ...       
Vol_4            US        560  N426US              117.0     ...       

       DepDelay  Origin  Dest  Distance CarrierDelay WeatherDelay  NASDelay  \
Vol_0       5.0     PBI   DCA       857            0            0         0   
Vol_1       5.0     PBI   DCA       857            0            0         0   
Vol_2       0.0     PBI   DCA       857            0            0         0   
Vol_3       0.0     PBI   DCA       857            0            0         0   
Vol_4      24.0     PBI   DCA       857            0            0         0   

       SecurityDelay  LateAircraftDelay  TotalDelay  
Vol_0              0                  0           0  
Vol_1              0                  0           0  
Vol_2              0                  0           0  
Vol_3              0                  0           0  
Vol_4              0                  0           0  

[5 rows x 23 columns]

In [58]:
df.tail()

DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime  \
Vol_301455          31          3     1838        1842     1943        1945   
Vol_301456          31          3      640         645     1307        1318   
Vol_301457          31          3     1222        1155     1422        1345   
Vol_301458          31          3     1830        1800     2057        2037   
Vol_301459          31          3     1615        1615     1755        1804   

           UniqueCarrier  FlightNum TailNum  ActualElapsedTime     ...      \
Vol_301455            DL        384  N997DL               65.0     ...       
Vol_301456            DL        385  N3745B              207.0     ...       
Vol_301457            DL        387  N917DL              120.0     ...       
Vol_301458            DL        389  N988DL              147.0     ...       
Vol_301459            DL        392   N6700              100.0     ...       

            DepDelay  Origin  Dest  Distance CarrierDelay WeatherDelay  \
Vol_301455      -4.0     SAV   ATL       215            0            0   
Vol_301456      -5.0     PHX   CVG      1569            0            0   
Vol_301457      27.0     BWI   ATL       576           27            0   
Vol_301458      30.0     BDL   ATL       859            0            0   
Vol_301459       0.0     ATL   FLL       581            0            0   

            NASDelay  SecurityDelay  LateAircraftDelay  TotalDelay  
Vol_301455         0              0                  0           0  
Vol_301456         0              0                  0           0  
Vol_301457        10              0                  0          37  
Vol_301458        17              0                  3          20  
Vol_301459         0              0                  0           0  

[5 rows x 23 columns]

### méthode 2

In [59]:
df['TotalDelay2'] = df[['CarrierDelay', 'WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay']].sum(axis=1)

In [60]:
df.head()

DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime  \
Vol_0           8          1      705         700      917         915   
Vol_1          10          3      705         700      920         915   
Vol_2          12          5      700         700      922         915   
Vol_3          14          7      700         700      910         915   
Vol_4          16          2      724         700      921         915   

      UniqueCarrier  FlightNum TailNum  ActualElapsedTime     ...       \
Vol_0            US        560  N405US              132.0     ...        
Vol_1            US        560  N784AU              135.0     ...        
Vol_2            US        560  N421US              142.0     ...        
Vol_3            US        560  N435US              130.0     ...        
Vol_4            US        560  N426US              117.0     ...        

       Origin  Dest  Distance  CarrierDelay WeatherDelay NASDelay  \
Vol_0     PBI   DCA       857             0            0        0   
Vol_1     PBI   DCA       857             0            0        0   
Vol_2     PBI   DCA       857             0            0        0   
Vol_3     PBI   DCA       857             0            0        0   
Vol_4     PBI   DCA       857             0            0        0   

       SecurityDelay  LateAircraftDelay  TotalDelay  TotalDelay2  
Vol_0              0                  0           0            0  
Vol_1              0                  0           0            0  
Vol_2              0                  0           0            0  
Vol_3              0                  0           0            0  
Vol_4              0                  0           0            0  

[5 rows x 24 columns]

In [61]:
df.tail()

DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime  \
Vol_301455          31          3     1838        1842     1943        1945   
Vol_301456          31          3      640         645     1307        1318   
Vol_301457          31          3     1222        1155     1422        1345   
Vol_301458          31          3     1830        1800     2057        2037   
Vol_301459          31          3     1615        1615     1755        1804   

           UniqueCarrier  FlightNum TailNum  ActualElapsedTime     ...       \
Vol_301455            DL        384  N997DL               65.0     ...        
Vol_301456            DL        385  N3745B              207.0     ...        
Vol_301457            DL        387  N917DL              120.0     ...        
Vol_301458            DL        389  N988DL              147.0     ...        
Vol_301459            DL        392   N6700              100.0     ...        

            Origin  Dest  Distance  CarrierDelay WeatherDelay NASDelay  \
Vol_301455     SAV   ATL       215             0            0        0   
Vol_301456     PHX   CVG      1569             0            0        0   
Vol_301457     BWI   ATL       576            27            0       10   
Vol_301458     BDL   ATL       859             0            0       17   
Vol_301459     ATL   FLL       581             0            0        0   

            SecurityDelay  LateAircraftDelay  TotalDelay  TotalDelay2  
Vol_301455              0                  0           0            0  
Vol_301456              0                  0           0            0  
Vol_301457              0                  0          37           37  
Vol_301458              0                  3          20           20  
Vol_301459              0                  0           0            0  

[5 rows x 24 columns]

### méthode 3

In [62]:
df['TotalDelay3'] = [x + y + z + t + u for x, y, z, t, u in zip(df.CarrierDelay, df.WeatherDelay, df.NASDelay, df.SecurityDelay, df.LateAircraftDelay)]

In [63]:
df.head()

DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime  \
Vol_0           8          1      705         700      917         915   
Vol_1          10          3      705         700      920         915   
Vol_2          12          5      700         700      922         915   
Vol_3          14          7      700         700      910         915   
Vol_4          16          2      724         700      921         915   

      UniqueCarrier  FlightNum TailNum  ActualElapsedTime     ...       Dest  \
Vol_0            US        560  N405US              132.0     ...        DCA   
Vol_1            US        560  N784AU              135.0     ...        DCA   
Vol_2            US        560  N421US              142.0     ...        DCA   
Vol_3            US        560  N435US              130.0     ...        DCA   
Vol_4            US        560  N426US              117.0     ...        DCA   

       Distance  CarrierDelay  WeatherDelay NASDelay SecurityDelay  \
Vol_0       857             0             0        0             0   
Vol_1       857             0             0        0             0   
Vol_2       857             0             0        0             0   
Vol_3       857             0             0        0             0   
Vol_4       857             0             0        0             0   

       LateAircraftDelay  TotalDelay  TotalDelay2  TotalDelay3  
Vol_0                  0           0            0            0  
Vol_1                  0           0            0            0  
Vol_2                  0           0            0            0  
Vol_3                  0           0            0            0  
Vol_4                  0           0            0            0  

[5 rows x 25 columns]

In [64]:
df.tail()

DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime  \
Vol_301455          31          3     1838        1842     1943        1945   
Vol_301456          31          3      640         645     1307        1318   
Vol_301457          31          3     1222        1155     1422        1345   
Vol_301458          31          3     1830        1800     2057        2037   
Vol_301459          31          3     1615        1615     1755        1804   

           UniqueCarrier  FlightNum TailNum  ActualElapsedTime     ...       \
Vol_301455            DL        384  N997DL               65.0     ...        
Vol_301456            DL        385  N3745B              207.0     ...        
Vol_301457            DL        387  N917DL              120.0     ...        
Vol_301458            DL        389  N988DL              147.0     ...        
Vol_301459            DL        392   N6700              100.0     ...        

            Dest  Distance  CarrierDelay  WeatherDelay NASDelay SecurityDelay  \
Vol_301455   ATL       215             0             0        0             0   
Vol_301456   CVG      1569             0             0        0             0   
Vol_301457   ATL       576            27             0       10             0   
Vol_301458   ATL       859             0             0       17             0   
Vol_301459   FLL       581             0             0        0             0   

            LateAircraftDelay  TotalDelay  TotalDelay2  TotalDelay3  
Vol_301455                  0           0            0            0  
Vol_301456                  0           0            0            0  
Vol_301457                  0          37           37           37  
Vol_301458                  3          20           20           20  
Vol_301459                  0           0            0            0  

[5 rows x 25 columns]

### Exercice : Chercher comment créer la colonne caractérisant si l'avion est en retard ou non (3 méthodes possibles)

### méthode 1

In [65]:
df['Delayed'] = [1 if x > 0 else 0 for x in df.TotalDelay]

In [66]:
df.head()

DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime  \
Vol_0           8          1      705         700      917         915   
Vol_1          10          3      705         700      920         915   
Vol_2          12          5      700         700      922         915   
Vol_3          14          7      700         700      910         915   
Vol_4          16          2      724         700      921         915   

      UniqueCarrier  FlightNum TailNum  ActualElapsedTime   ...     Distance  \
Vol_0            US        560  N405US              132.0   ...          857   
Vol_1            US        560  N784AU              135.0   ...          857   
Vol_2            US        560  N421US              142.0   ...          857   
Vol_3            US        560  N435US              130.0   ...          857   
Vol_4            US        560  N426US              117.0   ...          857   

       CarrierDelay  WeatherDelay  NASDelay SecurityDelay LateAircraftDelay  \
Vol_0             0             0         0             0                 0   
Vol_1             0             0         0             0                 0   
Vol_2             0             0         0             0                 0   
Vol_3             0             0         0             0                 0   
Vol_4             0             0         0             0                 0   

       TotalDelay  TotalDelay2  TotalDelay3  Delayed  
Vol_0           0            0            0        0  
Vol_1           0            0            0        0  
Vol_2           0            0            0        0  
Vol_3           0            0            0        0  
Vol_4           0            0            0        0  

[5 rows x 26 columns]

In [67]:
df.tail()

DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime  \
Vol_301455          31          3     1838        1842     1943        1945   
Vol_301456          31          3      640         645     1307        1318   
Vol_301457          31          3     1222        1155     1422        1345   
Vol_301458          31          3     1830        1800     2057        2037   
Vol_301459          31          3     1615        1615     1755        1804   

           UniqueCarrier  FlightNum TailNum  ActualElapsedTime   ...     \
Vol_301455            DL        384  N997DL               65.0   ...      
Vol_301456            DL        385  N3745B              207.0   ...      
Vol_301457            DL        387  N917DL              120.0   ...      
Vol_301458            DL        389  N988DL              147.0   ...      
Vol_301459            DL        392   N6700              100.0   ...      

            Distance  CarrierDelay  WeatherDelay  NASDelay SecurityDelay  \
Vol_301455       215             0             0         0             0   
Vol_301456      1569             0             0         0             0   
Vol_301457       576            27             0        10             0   
Vol_301458       859             0             0        17             0   
Vol_301459       581             0             0         0             0   

           LateAircraftDelay  TotalDelay  TotalDelay2  TotalDelay3  Delayed  
Vol_301455                 0           0            0            0        0  
Vol_301456                 0           0            0            0        0  
Vol_301457                 0          37           37           37        1  
Vol_301458                 3          20           20           20        1  
Vol_301459                 0           0            0            0        0  

[5 rows x 26 columns]

In [68]:
df.Delayed.dtype

dtype('int64')

In [69]:
df.Delayed.memory_usage() / 1024 / 1024

4.53680419921875

### méthode 2

In [70]:
df['Delayed2'] = df['TotalDelay'] > 0

In [71]:
df.head()

DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime  \
Vol_0           8          1      705         700      917         915   
Vol_1          10          3      705         700      920         915   
Vol_2          12          5      700         700      922         915   
Vol_3          14          7      700         700      910         915   
Vol_4          16          2      724         700      921         915   

      UniqueCarrier  FlightNum TailNum  ActualElapsedTime    ...     \
Vol_0            US        560  N405US              132.0    ...      
Vol_1            US        560  N784AU              135.0    ...      
Vol_2            US        560  N421US              142.0    ...      
Vol_3            US        560  N435US              130.0    ...      
Vol_4            US        560  N426US              117.0    ...      

       CarrierDelay  WeatherDelay  NASDelay  SecurityDelay LateAircraftDelay  \
Vol_0             0             0         0              0                 0   
Vol_1             0             0         0              0                 0   
Vol_2             0             0         0              0                 0   
Vol_3             0             0         0              0                 0   
Vol_4             0             0         0              0                 0   

      TotalDelay  TotalDelay2  TotalDelay3  Delayed  Delayed2  
Vol_0          0            0            0        0     False  
Vol_1          0            0            0        0     False  
Vol_2          0            0            0        0     False  
Vol_3          0            0            0        0     False  
Vol_4          0            0            0        0     False  

[5 rows x 27 columns]

In [72]:
df.tail()

DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime  \
Vol_301455          31          3     1838        1842     1943        1945   
Vol_301456          31          3      640         645     1307        1318   
Vol_301457          31          3     1222        1155     1422        1345   
Vol_301458          31          3     1830        1800     2057        2037   
Vol_301459          31          3     1615        1615     1755        1804   

           UniqueCarrier  FlightNum TailNum  ActualElapsedTime    ...     \
Vol_301455            DL        384  N997DL               65.0    ...      
Vol_301456            DL        385  N3745B              207.0    ...      
Vol_301457            DL        387  N917DL              120.0    ...      
Vol_301458            DL        389  N988DL              147.0    ...      
Vol_301459            DL        392   N6700              100.0    ...      

            CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
Vol_301455             0             0         0              0   
Vol_301456             0             0         0              0   
Vol_301457            27             0        10              0   
Vol_301458             0             0        17              0   
Vol_301459             0             0         0              0   

           LateAircraftDelay TotalDelay  TotalDelay2  TotalDelay3  Delayed  \
Vol_301455                 0          0            0            0        0   
Vol_301456                 0          0            0            0        0   
Vol_301457                 0         37           37           37        1   
Vol_301458                 3         20           20           20        1   
Vol_301459                 0          0            0            0        0   

            Delayed2  
Vol_301455     False  
Vol_301456     False  
Vol_301457      True  
Vol_301458      True  
Vol_301459     False  

[5 rows x 27 columns]

In [73]:
df.Delayed2.dtype

dtype('bool')

In [74]:
df.Delayed2.memory_usage() / 1024 / 1024

2.551952362060547

### méthode 3

In [75]:
df['Delayed3'] = df[['CarrierDelay', 'WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay']].any(axis=1)

In [76]:
df.head()

DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime  \
Vol_0           8          1      705         700      917         915   
Vol_1          10          3      705         700      920         915   
Vol_2          12          5      700         700      922         915   
Vol_3          14          7      700         700      910         915   
Vol_4          16          2      724         700      921         915   

      UniqueCarrier  FlightNum TailNum  ActualElapsedTime    ...     \
Vol_0            US        560  N405US              132.0    ...      
Vol_1            US        560  N784AU              135.0    ...      
Vol_2            US        560  N421US              142.0    ...      
Vol_3            US        560  N435US              130.0    ...      
Vol_4            US        560  N426US              117.0    ...      

       WeatherDelay  NASDelay  SecurityDelay  LateAircraftDelay TotalDelay  \
Vol_0             0         0              0                  0          0   
Vol_1             0         0              0                  0          0   
Vol_2             0         0              0                  0          0   
Vol_3             0         0              0                  0          0   
Vol_4             0         0              0                  0          0   

      TotalDelay2  TotalDelay3  Delayed  Delayed2  Delayed3  
Vol_0           0            0        0     False     False  
Vol_1           0            0        0     False     False  
Vol_2           0            0        0     False     False  
Vol_3           0            0        0     False     False  
Vol_4           0            0        0     False     False  

[5 rows x 28 columns]

In [77]:
df.tail()

DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime  \
Vol_301455          31          3     1838        1842     1943        1945   
Vol_301456          31          3      640         645     1307        1318   
Vol_301457          31          3     1222        1155     1422        1345   
Vol_301458          31          3     1830        1800     2057        2037   
Vol_301459          31          3     1615        1615     1755        1804   

           UniqueCarrier  FlightNum TailNum  ActualElapsedTime    ...     \
Vol_301455            DL        384  N997DL               65.0    ...      
Vol_301456            DL        385  N3745B              207.0    ...      
Vol_301457            DL        387  N917DL              120.0    ...      
Vol_301458            DL        389  N988DL              147.0    ...      
Vol_301459            DL        392   N6700              100.0    ...      

            WeatherDelay  NASDelay  SecurityDelay  LateAircraftDelay  \
Vol_301455             0         0              0                  0   
Vol_301456             0         0              0                  0   
Vol_301457             0        10              0                  0   
Vol_301458             0        17              0                  3   
Vol_301459             0         0              0                  0   

           TotalDelay TotalDelay2  TotalDelay3  Delayed  Delayed2  Delayed3  
Vol_301455          0           0            0        0     False     False  
Vol_301456          0           0            0        0     False     False  
Vol_301457         37          37           37        1      True      True  
Vol_301458         20          20           20        1      True      True  
Vol_301459          0           0            0        0     False     False  

[5 rows x 28 columns]

In [78]:
df.Delayed3.dtype

dtype('bool')

In [79]:
df.Delayed3.memory_usage() / 1024 / 1024

2.551952362060547

# Supprimer les colonnes inutiles parmi celles que l'on vient de créer
TotalDelay2, TotalDelay3, Delayed2 et Delayed3 sont inutiles car identiques à TotalDelay et Delayed

In [80]:
df.drop(['TotalDelay2', 'TotalDelay3', 'Delayed2', 'Delayed3'], axis=1, inplace=True)

In [81]:
df.head()

DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime  \
Vol_0           8          1      705         700      917         915   
Vol_1          10          3      705         700      920         915   
Vol_2          12          5      700         700      922         915   
Vol_3          14          7      700         700      910         915   
Vol_4          16          2      724         700      921         915   

      UniqueCarrier  FlightNum TailNum  ActualElapsedTime   ...     Origin  \
Vol_0            US        560  N405US              132.0   ...        PBI   
Vol_1            US        560  N784AU              135.0   ...        PBI   
Vol_2            US        560  N421US              142.0   ...        PBI   
Vol_3            US        560  N435US              130.0   ...        PBI   
Vol_4            US        560  N426US              117.0   ...        PBI   

       Dest  Distance  CarrierDelay WeatherDelay NASDelay  SecurityDelay  \
Vol_0   DCA       857             0            0        0              0   
Vol_1   DCA       857             0            0        0              0   
Vol_2   DCA       857             0            0        0              0   
Vol_3   DCA       857             0            0        0              0   
Vol_4   DCA       857             0            0        0              0   

       LateAircraftDelay  TotalDelay  Delayed  
Vol_0                  0           0        0  
Vol_1                  0           0        0  
Vol_2                  0           0        0  
Vol_3                  0           0        0  
Vol_4                  0           0        0  

[5 rows x 24 columns]

# Supprimer les différentes colonnes détaillant le retard
On ne s'intéresse qu'au retard total, on peut donc supprimer les colonnes détaillant les raisons du retard

In [82]:
df.drop(['CarrierDelay', 'WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay'], axis=1, inplace=True)

In [83]:
df.head()

DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime  \
Vol_0           8          1      705         700      917         915   
Vol_1          10          3      705         700      920         915   
Vol_2          12          5      700         700      922         915   
Vol_3          14          7      700         700      910         915   
Vol_4          16          2      724         700      921         915   

      UniqueCarrier  FlightNum TailNum  ActualElapsedTime  CRSElapsedTime  \
Vol_0            US        560  N405US              132.0           135.0   
Vol_1            US        560  N784AU              135.0           135.0   
Vol_2            US        560  N421US              142.0           135.0   
Vol_3            US        560  N435US              130.0           135.0   
Vol_4            US        560  N426US              117.0           135.0   

       AirTime  ArrDelay  DepDelay Origin Dest  Distance  TotalDelay  Delayed  
Vol_0    117.0       2.0       5.0    PBI  DCA       857           0        0  
Vol_1    120.0       5.0       5.0    PBI  DCA       857           0        0  
Vol_2    123.0       7.0       0.0    PBI  DCA       857           0        0  
Vol_3    120.0      -5.0       0.0    PBI  DCA       857           0        0  
Vol_4    105.0       6.0      24.0    PBI  DCA       857           0        0

# Modifier une colonne
Les colonnes DepTime, CRSDepTime, ArrTime et CRSArrTime sont formatées de la façon suivant : hhmm. Par exemple, un DepTime de 724 correspond à 7h24 le matin. On souhaite modifier cette colonne pour avoir le temps en minutes. Par exemple, 7h24 devient 444.

### Exercice : Créer une colonne DepTime2 qui passe l'information de DepTime en minutes

In [84]:
df['DepTime2'] = [60 * (x // 100) + x % 100 for x in df.DepTime]

In [85]:
df[['DepTime', 'DepTime2']].head()

DepTime  DepTime2
Vol_0      705       425
Vol_1      705       425
Vol_2      700       420
Vol_3      700       420
Vol_4      724       444

In [86]:
df[['DepTime', 'DepTime2']].tail()

DepTime  DepTime2
Vol_301455     1838      1118
Vol_301456      640       400
Vol_301457     1222       742
Vol_301458     1830      1110
Vol_301459     1615       975

### Exercice : Supprimer la colonne DepTime devenue obsolète

In [87]:
df.drop('DepTime', axis=1, inplace=True)

### Exercice : Renommer la colonne DepTime2 en DepTime (Je vous laisse chercher !)

In [95]:
df.rename(columns={'DepTime2' : 'DepTime'}, inplace=True)

### Exercice : Faire transformation pour CRSDepTime, ArrTime et CRSArrTime mais sans passer par une colonne intermédiaire

In [88]:
df['CRSDepTime'] = [60 * (x // 100) + x % 100 for x in df.CRSDepTime]

In [89]:
df['ArrTime'] = [60 * (x // 100) + x % 100 for x in df.ArrTime]

In [90]:
df['CRSArrTime'] = [60 * (x // 100) + x % 100 for x in df.CRSArrTime]

# Changer l'order des colonnes
La colonne DepTime2 a été créée en bout de DataFrame (tout à droite). Or, c'est celle que nous avons conservé en tant que DepTime suite à la suppression de la DepTime d'origne. On a donc changé l'ordre de nos colonnes

In [96]:
df.head()

DayofMonth  DayOfWeek  CRSDepTime  ArrTime  CRSArrTime UniqueCarrier  \
Vol_0           8          1         420      557         555            US   
Vol_1          10          3         420      560         555            US   
Vol_2          12          5         420      562         555            US   
Vol_3          14          7         420      550         555            US   
Vol_4          16          2         420      561         555            US   

       FlightNum TailNum  ActualElapsedTime  CRSElapsedTime  AirTime  \
Vol_0        560  N405US              132.0           135.0    117.0   
Vol_1        560  N784AU              135.0           135.0    120.0   
Vol_2        560  N421US              142.0           135.0    123.0   
Vol_3        560  N435US              130.0           135.0    120.0   
Vol_4        560  N426US              117.0           135.0    105.0   

       ArrDelay  DepDelay Origin Dest  Distance  TotalDelay  Delayed  DepTime  
Vol_0       2.0       5.0    PBI  DCA       857           0        0      425  
Vol_1       5.0       5.0    PBI  DCA       857           0        0      425  
Vol_2       7.0       0.0    PBI  DCA       857           0        0      420  
Vol_3      -5.0       0.0    PBI  DCA       857           0        0      420  
Vol_4       6.0      24.0    PBI  DCA       857           0        0      444

### Exercice : réordonner les colonnes pour revenir à l'ordre d'origine

In [97]:
cols = df.columns.tolist()[0:2] + ['DepTime'] + df.columns.tolist()[2:-1]

In [98]:
cols

['DayofMonth',
 'DayOfWeek',
 'CRSDepTime',
 'DepTime',
 'ArrTime',
 'CRSArrTime',
 'UniqueCarrier',
 'FlightNum',
 'TailNum',
 'ActualElapsedTime',
 'CRSElapsedTime',
 'AirTime',
 'ArrDelay',
 'DepDelay',
 'Origin',
 'Dest',
 'Distance',
 'TotalDelay',
 'Delayed']

In [99]:
df = df[cols]

In [100]:
df.head()

DayofMonth  DayOfWeek  CRSDepTime  DepTime  ArrTime  CRSArrTime  \
Vol_0           8          1         420      425      557         555   
Vol_1          10          3         420      425      560         555   
Vol_2          12          5         420      420      562         555   
Vol_3          14          7         420      420      550         555   
Vol_4          16          2         420      444      561         555   

      UniqueCarrier  FlightNum TailNum  ActualElapsedTime  CRSElapsedTime  \
Vol_0            US        560  N405US              132.0           135.0   
Vol_1            US        560  N784AU              135.0           135.0   
Vol_2            US        560  N421US              142.0           135.0   
Vol_3            US        560  N435US              130.0           135.0   
Vol_4            US        560  N426US              117.0           135.0   

       AirTime  ArrDelay  DepDelay Origin Dest  Distance  TotalDelay  Delayed  
Vol_0    117.0       2.0       5.0    PBI  DCA       857           0        0  
Vol_1    120.0       5.0       5.0    PBI  DCA       857           0        0  
Vol_2    123.0       7.0       0.0    PBI  DCA       857           0        0  
Vol_3    120.0      -5.0       0.0    PBI  DCA       857           0        0  
Vol_4    105.0       6.0      24.0    PBI  DCA       857           0        0

# Synthèse des retards par compagnie
La méthode groupby va nous permettre d'extraire facilement des informations par compagnie : https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.groupby.html

### Création d'un nouveau DataFrame

In [101]:
df_comp = pd.DataFrame(index=df.UniqueCarrier.unique())

In [102]:
df_comp

Empty DataFrame
Columns: []
Index: [US, WN, YV, OH, OO, XE, TZ, UA, DL, EV, F9, FL, HA, MQ, NW, AA, AQ, AS, B6, CO]

In [103]:
df_comp.sort_index(inplace=True)

In [104]:
df_comp.head()

Empty DataFrame
Columns: []
Index: [AA, AQ, AS, B6, CO]

### Première colonne du DataFrame

In [105]:
df_comp['count'] = df.loc[:, ['UniqueCarrier', 'TotalDelay']].groupby('UniqueCarrier').count()

In [106]:
df_comp.head()

count
AA  27349
AQ   1800
AS   6688
B6   5985
CO  12904

### Exercice : Continuer à remplir le DataFrame avec la moyenne, l'écart-type, le minimum, le quantile 25%, le quantile 50%, le quantile 75% et le maximum

In [107]:
df_comp['moyenne'] = df.loc[:, ['UniqueCarrier', 'TotalDelay']].groupby('UniqueCarrier').mean()

In [108]:
df_comp['ecartType'] = df.loc[:, ['UniqueCarrier', 'TotalDelay']].groupby('UniqueCarrier').std()

In [109]:
df_comp['minimum'] = df.loc[:, ['UniqueCarrier', 'TotalDelay']].groupby('UniqueCarrier').min()

In [110]:
df_comp['quantile25'] = df.loc[:, ['UniqueCarrier', 'TotalDelay']].groupby('UniqueCarrier').quantile(0.25)

In [111]:
df_comp['quantile50'] = df.loc[:, ['UniqueCarrier', 'TotalDelay']].groupby('UniqueCarrier').quantile(0.5)

In [112]:
df_comp['quantile75'] = df.loc[:, ['UniqueCarrier', 'TotalDelay']].groupby('UniqueCarrier').quantile(0.75)

In [113]:
df_comp['maximum'] = df.loc[:, ['UniqueCarrier', 'TotalDelay']].groupby('UniqueCarrier').max()

In [114]:
df_comp.head()

count    moyenne  ecartType  minimum  quantile25  quantile50  quantile75  \
AA  27349  13.237486  36.670329        0         0.0         0.0         0.0   
AQ   1800   3.396111  13.197027        0         0.0         0.0         0.0   
AS   6688   7.711872  26.100351        0         0.0         0.0         0.0   
B6   5985   8.839599  28.645614        0         0.0         0.0         0.0   
CO  12904  13.964585  35.367171        0         0.0         0.0        15.0   

    maximum  
AA      993  
AQ      235  
AS      523  
B6      653  
CO      592

# Opérations plus complexes dans un groupby

### Application de la fonction finale à plusieurs colonnes

In [117]:
df.loc[:, ['UniqueCarrier', 'TotalDelay', 'CRSElapsedTime']].groupby('UniqueCarrier').max()

TotalDelay  CRSElapsedTime
UniqueCarrier                            
AA                    993           549.0
AQ                    235           354.0
AS                    523           414.0
B6                    653           400.0
CO                    592           640.0
DL                    560           579.0
EV                   1061           226.0
F9                    471           325.0
FL                    561           335.0
HA                    283           385.0
MQ                    825           209.0
NW                   1078           495.0
OH                    535           196.0
OO                    462           236.0
TZ                    293           392.0
UA                    526           561.0
US                    622           425.0
WN                    410           375.0
XE                    489           239.0
YV                    467           232.0

### grouper selon plusieurs colonnes

In [116]:
df.loc[:, ['Origin', 'Dest', 'TotalDelay']].groupby(['Origin', 'Dest']).max()

TotalDelay
Origin Dest            
ABE    ATL          394
       CLE           60
       CLT            0
       CVG            0
       JFK            0
       ORD          200
ABI    DFW           87
ABQ    AMA           80
       ATL           98
       BWI            0
       CVG           27
       DAL           52
       DEN          164
       DFW          614
       ELP          198
       HOU           26
       IAH          215
       LAS          141
       LAX           35
       LBB           58
       MAF           87
       MCI           35
       MCO          112
       MDW           80
       MSP           27
       OAK           50
       ORD          100
       PDX            0
       PHX           74
       SAN           70
...                 ...
TYS    CLE          243
       CVG           67
       DEN            0
       DFW          102
       EWR           94
       IAH          241
       LGA           18
       ORD          406
VLD    ATL           94
VPS    ATL          216
       DFW          582
       IAH          125
       MEM           77
WRG    KTN           36
       PSG           16
XNA    ATL          201
       CVG            0
       DEN           45
       DFW          480
       EWR          286
       IAH          144
       LAX          243
       LGA          127
       ORD          323
       SLC           57
YAK    CDV           22
       JNU            0
YUM    IPL           55
       LAX          100
       PHX          219

[4003 rows x 1 columns]

### Utilisation d'une lambda fonction

In [118]:
df.loc[:, ['UniqueCarrier', 'TotalDelay', 'CRSElapsedTime']].groupby('UniqueCarrier').apply(lambda x: 100*(x.TotalDelay / x.CRSElapsedTime).max())

UniqueCarrier
AA     844.827586
AQ     382.758621
AS     462.831858
B6     932.857143
CO     442.307692
DL     708.860759
EV    1583.582090
F9     336.428571
FL     623.333333
HA     282.758621
MQ    1015.942029
NW    1847.169811
OH     983.333333
OO    1320.000000
TZ     237.500000
UA     634.146341
US     659.375000
WN     611.111111
XE     738.333333
YV     603.703704
dtype: float64

### Exercice : Certains retards semblent énormes (+ de 1000% !). Isolez les lignes correspondant à ces vols et observez ce qu'il s'est passé

In [119]:
indexes = df.loc[:, ['UniqueCarrier', 'TotalDelay', 'CRSElapsedTime']].groupby('UniqueCarrier').apply(lambda x: (x.TotalDelay / x.CRSElapsedTime).idxmax())

In [120]:
df.loc[indexes]

DayofMonth  DayOfWeek  CRSDepTime  DepTime  ArrTime  CRSArrTime  \
Vol_235600          13          6         582     1055     1130         640   
Vol_253989           2          2         600      714      740         629   
Vol_261259          27          6         525     1056     1221         698   
Vol_262694           3          3        1360      566      643        1430   
Vol_280808          14          7         930     1279     1353        1008   
Vol_297807          25          4         990       85      129        1009   
Vol_162756          16          2         360     1430     1548         487   
Vol_165813          23          2         380      861     1051         580   
Vol_171396          14          7         420      973     1071         510   
Vol_178364          27          6        1120     1119     1231        1149   
Vol_185630          29          1        1180      445      510        1249   
Vol_214261          30          2         900      419      492         953   
Vol_76366           17          3         440      911     1020         548   
Vol_103042          29          1         525      989     1022         560   
Vol_124035          31          3         840     1095     1185         900   
Vol_130212          24          3         978     1236       80        1000   
Vol_9515             5          5         420      793      906         484   
Vol_52881           25          4        1205       35       85        1250   
Vol_110113           6          6         350      785      853         410   
Vol_72176           24          3        1225      113      165        1279   

           UniqueCarrier  FlightNum TailNum  ActualElapsedTime  \
Vol_235600            AA        937  N069AA               75.0   
Vol_253989            AQ         72  N821AL               26.0   
Vol_261259            AS         60  N782AS              105.0   
Vol_262694            B6        128  N580JB               77.0   
Vol_280808            CO        323  N79402               74.0   
Vol_297807            DL       1226  N907DL              104.0   
Vol_162756            EV       4900  N901EV               58.0   
Vol_165813            F9        420  N912FR              130.0   
Vol_171396            FL        120  N281AT               98.0   
Vol_178364            HA        314  N478HA              112.0   
Vol_185630            MQ       3543  N501BG               65.0   
Vol_214261            NW       1098  N771NC               73.0   
Vol_76366             OH       5319  N331CA               49.0   
Vol_103042            OO       6225  N294SW               33.0   
Vol_124035            TZ       4293  N514AT              150.0   
Vol_130212            UA        371  N813UA              344.0   
Vol_9515              US          3  N621AW              113.0   
Vol_52881             WN        183    N754               50.0   
Vol_110113            XE       3069  N16944               68.0   
Vol_72176             YV       7309  N75996               52.0   

            CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin Dest  Distance  \
Vol_235600            58.0     42.0     490.0     473.0    MCO  MIA       193   
Vol_253989            29.0     19.0     111.0     114.0    LIH  HNL       102   
Vol_261259           113.0     91.0     523.0     531.0    KTN  SEA       680   
Vol_262694            70.0     45.0     653.0     646.0    JFK  BTV       267   
Vol_280808            78.0     41.0     345.0     349.0    IAH  MSY       305   
Vol_297807            79.0     89.0     560.0     535.0    CVG  ORD       264   
Vol_162756            67.0  -1402.0    1061.0    1070.0    MEI  ATL       267   
Vol_165813           140.0    113.0     471.0     481.0    SAN  DEN       853   
Vol_171396            90.0     74.0     561.0     553.0    TPA  ATL       406   
Vol_178364            29.0     96.0      82.0      -1.0    LIH  HNL       102   
Vol_185630            69.0     49.0     701.0     705.0    DFW  LIT       

# Sauvegarde des données
la méthode to_csv permet d'écrire le DataFrame modifié au format csv. L'argument index=True précise que l'on souhaite écrire l'index dans le fichier

In [121]:
df.to_csv('datas_cleaned.csv', index=True)